[ty] Initial implementation of declaration and definition providers.

UnboundVariable opened the pull request:
This PR implements "go to definition" and "go to declaration" functionality for name nodes only. Future PRs will add support for attributes, module names in import statements, keyword argument names, etc.

This PR:
* Registers a declaration and definition request handler for the language server.
* Splits out the `goto_type_definition` into its own module. The `goto` module contains functionality that is common to `goto_type_definition`, `goto_declaration` and `goto_definition`.
* Roughs in a new module `stub_mapping` that is not yet implemented. It will be responsible for mapping a definition in a stub file to its corresponding definition(s) in an implementation (source) file.
* Adds a new IDE support function `definitions_for_name` that collects all of the definitions associated with a name and resolves any imports (recursively) to find the original definitions associated with that name.
* Adds a new `VisibleAncestorsIter` stuct that iterates up the scope hierarchy but skips scopes that are not visible to starting scope.

AlexWaygood commented:
> Merging #19371 will **degrade performances by 5.26%**

I think that's just a race condition due to https://github.com/astral-sh/ruff/pull/19328 being merged at exactly the wrong time -- if you rebase this PR, it should go away

UnboundVariable commented:
To potential reviewers... Most of the changes in this PR are straightforward and probably not worth your time reviewing in any great detail. Here are the specific areas where I would most appreciate review and feedback:

1. In [semantic_index.rs](https://github.com/astral-sh/ruff/pull/19371/files#diff-2c8485d2796b0e1e643d3e48cd572c9810dcc7f45bbecae951ca00e5276d7d9d), I added a `VisibleAncestorsIter` struct to walk up the scope hierarchy but skip the scopes that are not visible to the starting scope. I would have expected to see functionality like this already, but I couldn't find it. Let me know if I should be using something else here.
2. I added a function [definitions_for_name](https://github.com/astral-sh/ruff/pull/19371/files#diff-f9eb3d520f255c7ff2246e1250a1ee679eea2e642bef437c28cc36f49e797cee) in `ide_support.rs`. Does this look like the right abstraction and separation of concern? I tried a couple of different options, but some of my other attempts required exposing way more public structs and methods from `ty_python_semantic`  to the `ty_ide` crate. 
3. I added a new module [resolve_definition](https://github.com/astral-sh/ruff/pull/19371/files#diff-940d4c36a4b3d1411e05f62bcbf93457712b47d29e5afc2bcbb751c51286ad1d) that is used by `definitions_for_name`. It contains code that "resolves" import declarations -- following them recursively to locate the original source declaration. Does this look like the right approach?

MichaReiser reviewed (commented):
Thank. I don't have much to add other than I agree with @carljm's summary: It be curious to hear your thoughts on duplicating some of the semantic behavior for the LSP vs trying to enrhich `ty_python_semantic` itself to avoid the code duplication.

carljm commented:
@UnboundVariable and I discussed this today in person and felt that it makes sense to continue with the current approach, even though it involves some of what looks like duplication, because in fact the duplication is "shallower" than it appears; there are a lot of behavior differences between type inference and collecting definitions for goto-definition. (One upcoming example with attributes is that goto-definition actually shouldn't run the descriptor protocol, which is a big part of our type-inference attribute-access code.) Having to integrate all of these differences into the type-inference code as optional flags will increase complexity and slow down implementing the LSP features, as well as making it harder to make future changes to the behavior. In other words, we feel this is a case where it makes sense to accept some duplication in exchange for greater flexibility.

UnboundVariable commented on crates/ty_ide/src/goto.rs:
```diff
@@ -154,6 +136,101 @@ impl GotoTarget<'_> {
+
+    /// Get navigation targets for definitions associated with a name expression
+    fn get_name_definition_targets(
+        _expression: ruff_python_ast::ExprRef<'_>,
```
Good catch. That was left over from an earlier iteration of the code.

UnboundVariable commented on crates/ty_python_semantic/src/types/ide_support.rs:
```diff
@@ -386,6 +388,121 @@ pub fn definition_kind_for_name<'db>(
+            break;
+        }
+
+        // If marked as nonlocal, skip current scope and search in ancestor scopes
```
With language server functionality like "go to declaration", it's not clear what behavior is "correct". It's somewhat subjective, and we will undoubtedly need to iterate on all of the language server features in response to user feedback. As you and I discussed in person, my recommended approach is to use Microsoft's language server (pylance) as a starting point to inform the initial behavior. Pylance has already gone through several years of iteration and user feedback, so it's probably a good starting point for ty.

Another general principle with "go to declaration" and "go to definition" is that it's probably not a good idea to return different results for different usages of a symbol. In other words, if you "go to definition" from any use of a symbol, it should display the same list of definitions.

Based on these principles, the "correct" solution here is to address the TODO that I left in the code. Once we do that, we will want to skip the current scope here. In other words, I wrote the code assuming that we will eventually fix the TODO. If you have strong opinions about how this should work in the interim — or prefer that I prioritize fixing the TODO now, let me know.
